File: arrays_and_strings/0977_square_of_a_sorted_array/python/solution.py


from typing import List, Optional, Dict, Set

#  Squares of a Sorted Array

# LeetCode: https://leetcode.com/problems/-squares-of-a-sorted-array/

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        i,j = 0,n-1       # initialized pointers 
        result = [0]*n    # initialized result arrays for easy indexing  

        for k in range(n-1,-1,-1):  #k starts at n-1 decreasing till 0
            
            if nums[i]**2 > nums[j]**2:
                result[k] = nums[i]**2
                i += 1
            else:
                result[k] = nums[j]**2
                j -= 1
        return result     
        
# --- Driver Code for Local Testing ---

if __name__ == "__main__":
    sol = Solution()
    result = sol.sortedSquares([-4,-3,1, 2, 3])
    print(f"Result: {result}")

File: arrays_and_strings/0977_square_of_a_sorted_array/rust/solution.rs
// My first rust solution

#[derive(Debug)]
struct Solution;

impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let n:usize = nums.len();
        // initializing pointers 
        let mut i:usize = 0;
        let mut j:usize = n-1;

        // initializing result array
        let mut result:Vec<i32> = vec![0;n];
        
        // k starts at n-1 decreasing till 0
        for k in (0..n).rev(){
            if nums[i].pow(2) > nums[j].pow(2){
                result[k] = nums[i].pow(2);
                i += 1;
            }
            else{
                result[k] = nums[j].pow(2);
                j -= 1;
            }

        }
        return result 
    }
}

fn main() {
    //  Define sample input vector
    let nums = vec![-4, -1, 0, 3, 10];
    
    // Call the function
    let output = Solution::sorted_squares(nums);
    
    // Print the result using debug formatting 
    println!("Sorted Squares: {output:?}");
}